python/ray/data/dataset.py

```diff
@@ -2261,6 +2262,70 @@ def write_csv(
             **arrow_csv_args,
         )
 
+    def write_tfrecords(
+        self,
+        path: str,
+        *,
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        try_create_dir: bool = True,
+        arrow_open_stream_args: Optional[Dict[str, Any]] = None,
+        block_path_provider: BlockWritePathProvider = DefaultBlockWritePathProvider(),
+        ray_remote_args: Dict[str, Any] = None,
+    ) -> None:
+        """Write the dataset to TFRecord files.
+
+        The `TFRecord <https://www.tensorflow.org/tutorials/load_data/tfrecord>`_
+        files will contain
+        `tf.train.Example <https://www.tensorflow.org/api_docs/python/tf/train/Example>`_ # noqa: E501
+        records, with one Example record for each row in the dataset.
+
+        .. warning::
+            tf.train.Feature only natively stores ints, floats, and bytes,
+            so this function only supports datasets with these data types,
+            and will error if the dataset contains unsupported types.
+
+        This is only supported for datasets convertible to Arrow records.
+        To control the number of files, use ``.repartition()``.
+
+        Unless a custom block path provider is given, the format of the output
+        files will be {uuid}_{block_idx}.tfrecords, where ``uuid`` is an unique id
+        for the dataset.
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.from_items([
+            ...     { "name": "foo", "score": 42 },
+            ...     { "name": "bar", "score": 43 },
+            ... ])
+            >>> ds.write_tfrecords("s3://bucket/path") # doctest: +SKIP
+
+        Time complexity: O(dataset size / parallelism)
+
+        Args:
+            path: The path to the destination root directory, where tfrecords
+                files will be written to.
+            filesystem: The filesystem implementation to write to.
```

Comment:
The `filesystem` is null-able. Can you document what's the behavior if it's None?

Reply:
Let me investigate what the behaviour would be. Currently all the write_* methods have this same signature and docstring for `filesystem`.

Reply:
Looks like all these file based datasources will [resolve it using](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/file_based_datasource.py#L605) `pyarrow.fs._resolve_filesystem_and_path`. 

I don't have the full context to understand under what conditions a None filesystem is resolvable here - @clarkzinzow would you know?
If you'd like, I can add a comment to write_csv, write_parquet, write_json, write_numpy, write_tfrecorsd that a None filesystem will be resolved by `pyarrow.fs`.

Reply:
Yeah, it's inferred from provided files.

Right, this seems a stretch since all other APIs do not have a mention about what to expect if it's None. Thank you Charles!